Implement more support for per-channel quantization

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26339 Serialization for per channel qtensor
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* #26242 Fold activation permutation inside quantized conv operator
* #26241 Fold weight permutation inside quantized conv operator
* **#26240 Implement more support for per-channel quantization**

In particular adds support for empty/empty_like which is needed for memory layouts to work.

Differential Revision: [D17443220](https://our.internmc.facebook.com/intern/diff/D17443220)

Commit:
implement more per-channel code

jerryzh168 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -217,14 +217,24 @@ Tensor empty_like(
-                                       use_memory_format);
+    auto qscheme = self.qscheme();
+    if (qscheme == kPerTensorAffine) {
+      return at::_empty_affine_quantized(self.sizes(), options,
```
we have different ordering of arguments here...
maybe we can have:
```at::_empty_affine_quantized(q_scale, q_zero_point, sizes, options, use_memory_format);```
and
```at::_empty_per_channel_affine_quantized_like(q_per_channel_scales, q_per_channel_zero_points, q_per_channel_axis, sizes, options, use_memory_format);```

But should do in a separate PR: @dskhudia could you change the order for `_empty_per_channel_affine_quantized_like`?

jerryzh168 replied:
oh `q_per_channel_axis` is added in this PR, then @dzhulgakov could you change the order as well?

jerryzh168 replied:
looks like this is addressed in the later PRs

dzhulgakov replied:
So the order and name for `_empty_per_channel_affine_quantized_like` (_like!) - was weird as a workaround for some codegen peculiarity - https://github.com/pytorch/pytorch/pull/26243 fixes it

I the order is still not the same - TensorOptions and additional arguments are in the wrong order. It's C++ only as the args are marked as kwargs in python. But we should fix it anyway. Since it's an intrusive change - I'll send a separate PR for it after these ones land

jerryzh168 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -33,7 +33,7 @@
     'numpy_T',  # this needs to be an attribute in Python, not a function
     'nonzero(_(out|numpy))?',
-    'set_quantizer_',
+    'set_quantizer_', 'q_per_channel_axis',  # return types not supported yet
```
why? we probably need this for serializing PerChannelAffine tensors

jerryzh168 replied:
do you mean quantizer here?

dzhulgakov replied:
apparently we don't support returning int[] :-\ I can look into fixing it properly

dzhulgakov replied:
ah, I was just lazy - it was easy enough to patch

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

[ghstack-poisoned]

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
+        # change memory format
+        qlast = qr.contiguous(memory_format=torch.channels_last)
+        self.assertEqual(qr.stride(), list(reversed(sorted(qr.stride()))))
+        self.assertNotEqual(qlast.stride(), list(reversed(sorted(qlast.stride()))))
```
what is this for? to check the contiguous operation actually happened?

dzhulgakov replied:
yeah, just making sure. I can remove it, but it doesn't hurt

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
+        self.assertTrue(torch.equal(qr.int_repr(), qlast.int_repr()))
+        self.assertTrue(qr.q_scale() == qlast.q_scale())
+        self.assertTrue(qr.q_zero_point() == qlast.q_zero_point())
+        self.assertTrue(np.array_equal(qlast.dequantize().numpy(), qr.dequantize().numpy()))
```
I think we have `self.assertEqual` on pytorch tensor as well. so
```
self.assertEqual(qlast.dequantize(), qr.dequantize())
```
should work

dzhulgakov replied:
ah, good point! when I tried it I typed assertEquals instead of assertEqual and we don't override that one :-P

jerryzh168 commented on test/test_quantized_tensor.py:
```diff
@@ -172,10 +174,40 @@ def test_qtensor_permute(self):
+        qlast = qr.contiguous(memory_format=torch.channels_last)
+        self.assertEqual(qr.stride(), list(reversed(sorted(qr.stride()))))
+        self.assertNotEqual(qlast.stride(), list(reversed(sorted(qlast.stride()))))
+        self.assertTrue(torch.equal(qr.int_repr(), qlast.int_repr()))
```
FYI: int_repr() will do contiguous right now, but I think it should preserve the strides,  will fix in https://github.com/pytorch/pytorch/pull/25429 when I have some time.

jerryzh168 reviewed (approved):
LGTM Thanks

jerryzh168 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -347,7 +347,13 @@ Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef s
-  auto result = detail::make_tensor<QTensorImpl>(Storage(self.storage()), self.type_set(), get_qtensorimpl(self)->quantizer());
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(
+      quantizer->qscheme() == QScheme::PER_TENSOR_SYMMETRIC ||
```
we don't have `QScheme::PER_TENSOR_SYMMETRIC` in backend. please remove.

see: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/quantized/Quantizer.h#L121

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

[ghstack-poisoned]

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -218,14 +218,24 @@ Tensor empty_like(
+                                         self.q_scale(),
+                                         self.q_zero_point(),
+                                         use_memory_format);
+    } else if (qscheme == kPerChannelAffine) {
```
Do we also need to support kPerChannelSymmetric and kPerTensorSymmetric here?

dzhulgakov replied:
again, they are not present in the backend

raghuramank100 commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -348,7 +348,12 @@ Tensor as_strided_tensorimpl(const Tensor& self, IntArrayRef size, IntArrayRef s
-  auto result = detail::make_tensor<QTensorImpl>(Storage(self.storage()), self.type_set(), get_qtensorimpl(self)->quantizer());
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(
+      quantizer->qscheme() == QScheme::PER_TENSOR_AFFINE,
```
Similar comments as above for other qschemes

raghuramank100 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -94,8 +94,10 @@ Tensor q_per_channel_zero_points_quant(const Tensor& self) {
-  return get_qtensorimpl(self)->quantizer().get();
+IntArrayRef q_per_channel_axis_quant(const Tensor& self) {
+  auto quantizer = get_qtensorimpl(self)->quantizer();
+  TORCH_CHECK(quantizer->qscheme() == kPerChannelAffine);
```
Support for PerChannelSymmetric

raghuramank100 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -183,6 +185,10 @@ Tensor& set_quantizer_(Tensor& self, ConstQuantizerPtr quantizer) {
 
 Tensor quantized_clone(const Tensor& self) {
+  // TODO: add per channel support
+  TORCH_INTERNAL_ASSERT(
```
Similar comments as above

raghuramank100 commented on test/test_quantized_tensor.py:
```diff
@@ -77,13 +77,15 @@ def test_per_channel_qtensor_creation(self):
         q = torch._empty_per_channel_affine_quantized_like(scales, zero_points, [numel], [ch_axis], dtype=torch.quint8)
         self.assertEqual(scales, q.q_per_channel_scales())
         self.assertEqual(zero_points, q.q_per_channel_zero_points())
+        self.assertEqual([ch_axis], q.q_per_channel_axis())
```
Both per_channel and qtensor creation tests should sweep over symmetric and affine quant.

raghuramank100 commented on test/test_quantized_tensor.py:
```diff
@@ -153,28 +155,58 @@ def quantize_c(data, scales, zero_points):
+        self.assertEqual(qlast.dequantize(), qr.dequantize())
+
+    def test_qtensor_per_channel_permute(self):
+        r = torch.rand(20, 10, 2, 2, dtype=torch.float) * 4 - 2
```
Can we allow for permutes that do not change the axis that is quantized?
Would this work?
```python
# x is in nchw
i.e x_q = torch.quantize_linear_per_channel(... axis =[0])
# x_q is also in nchw
x_nhwc =  x_q.permute([0, 2, 3,1])
```

dzhulgakov replied:
as I mentioned in another place, we could, but I think we can add it later in a separate diff. And memory format works today already so it should be sufficient

raghuramank100 commented on torch/_tensor_docs.py:
```diff
@@ -1903,15 +1903,27 @@ def callable(a, b) -> number
+q_per_channel_zero_points() -> tuple of ints
+
+Given a Tensor quantized by linear (affine) per-channel quantization,
+returns a indices of dimensions on which per-channel quantization is applied.
```
Do we really support a tuple for axis? Why not restrict axis to be scalar/list of length 1 and clarify this in the API.
Also NIT: the doc string  should say q_per_channel_axis()

dzhulgakov replied:
How likely are we to have multidim in the future? Having both apis would be confusing

raghuramank100 reviewed (changes requested):
Please see comments, thanks

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

Differential Revision: [D17443220](https://our.internmc.facebook.com/intern/diff/D17443220)

[ghstack-poisoned]

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

Differential Revision: [D17443220](https://our.internmc.facebook.com/intern/diff/D17443220)

[ghstack-poisoned]

Commit:
Update on "Implement more support for per-channel quantization"


In particular adds support for empty/empty_like which is needed for memory layouts to work.

Differential Revision: [D17443220](https://our.internmc.facebook.com/intern/diff/D17443220)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -94,8 +94,10 @@ Tensor q_per_channel_zero_points_quant(const Tensor& self) {
 
-Quantizer* quantizer(const Tensor& self) {
-  return get_qtensorimpl(self)->quantizer().get();
+IntArrayRef q_per_channel_axis_quant(const Tensor& self) {
```
This is the very first occurrence of an `IntArrayRef` at return site of a native function, and I don't like it. The returned ref is non-owning. So what's its lifetime tied to? This is not documented anywhere. (In fact, the lifetime is tied to the *quantizer*, not the tensor itself; so the semantics don't even match `sizes()` and `strides()`, whose lifetime is tied to the tensor). Furthermore, this would have to be supported properly in the JIT, as all values in the stack must be owning. C.f. implementation of `sizes()`:

```
    Operator(
        "aten::size(Tensor self) -> int[]",
        [](Stack& stack) {
          RECORD_FUNCTION("size", last(stack, 1));
            
          auto t = std::move(pop(stack)).toTensor();
          pack(stack, t.sizes().vec());
          return 0;
        },    
        aliasAnalysisFromSchema()),
```

I suppose, from first principles, the ability to return a primitive array from a function is desirable functionality, I think we should think carefully about how we want to design this. As a first proposal, the return type of this function should be a `std::vector<>`, not an `IntArrayRef`.

cc @gchanan @zdevito
